cdap-app-fabric/src/test/java/io/cdap/cdap/internal/tethering/TetheringClientHandlerTest.java

```diff
@@ -476,6 +478,87 @@ public void testNullEndpoint() throws Exception {
     Assert.assertEquals(HttpResponseStatus.BAD_REQUEST.code(), response.getResponseCode());
   }
 
+  @Test
+  public void testServerReturns404ForConnectedPeer() throws Exception {
+    // Server returns 404 before tethering has been accepted or rejected by admin on server side.
+    serverHandler.setResponseStatus(HttpResponseStatus.NOT_FOUND, NOT_FOUND_MESSAGE);
+
+    // Client initiates tethering with the server
+    createTethering(SERVER_INSTANCE, PROJECT, LOCATION, NAMESPACES, REQUEST_TIME, DESCRIPTION);
+
+    // Server accepts tethering
+    serverHandler.setResponseStatus(HttpResponseStatus.OK);
+
+    String serverEndpoint = serverConfig.getConnectionConfig().getURI().toString();
+    // Tethering state should transition to accepted on the client.
+    waitForTetheringStatus(TetheringStatus.ACCEPTED, SERVER_INSTANCE, serverEndpoint,
+                           PROJECT, LOCATION, NAMESPACES, DESCRIPTION, true);
+
+    // Server return NOT_FOUND on the next poll
+    serverHandler.setResponseStatus(HttpResponseStatus.NOT_FOUND, NOT_FOUND_MESSAGE);
+
+    // Tethering state should transition to rejected on the client.
+    waitForTetheringStatus(TetheringStatus.REJECTED, SERVER_INSTANCE, serverEndpoint,
+                           PROJECT, LOCATION, NAMESPACES, DESCRIPTION, true);
+
+    // cleanup
+    deleteTethering(SERVER_INSTANCE);
```

Comment:
cleanup that always needs to be performed is best done in a `@After` annotated method. Otherwise a failure earlier in this method will cause this cleanup to be skipped, which can impact other tests.

Reply:
done